src/lattice.cpp

```diff
@@ -0,0 +1,945 @@
+  to_hdf5_inner(lat_group);
+
+  close_group(lat_group);
+}
+
+//==============================================================================
+// RectLattice implementation
+//==============================================================================
+
+RectLattice::RectLattice(pugi::xml_node lat_node)
+  : Lattice {lat_node}
+{
+  // Read the number of lattice cells in each dimension.
+  std::string dimension_str {get_node_value(lat_node, "dimension")};
+  std::vector<std::string> dimension_words {split(dimension_str)};
+  if (dimension_words.size() == 2) {
+    n_cells[0] = stoi(dimension_words[0]);
+    n_cells[1] = stoi(dimension_words[1]);
+    n_cells[2] = 1;
+    is_3d = false;
+  } else if (dimension_words.size() == 3) {
+    n_cells[0] = stoi(dimension_words[0]);
+    n_cells[1] = stoi(dimension_words[1]);
+    n_cells[2] = stoi(dimension_words[2]);
+    is_3d = true;
+  } else {
+    fatal_error("Rectangular lattice must be two or three dimensions.");
+  }
+
+  // Read the lattice lower-left location.
+  std::string ll_str {get_node_value(lat_node, "lower_left")};
+  std::vector<std::string> ll_words {split(ll_str)};
+  if (ll_words.size() != dimension_words.size()) {
+    fatal_error("Number of entries on <lower_left> must be the same as the "
+                "number of entries on <dimension>.");
+  }
+  lower_left[0] = stod(ll_words[0]);
+  lower_left[1] = stod(ll_words[1]);
+  if (is_3d) {lower_left[2] = stod(ll_words[2]);}
+
+  // Read the lattice pitches.
+  std::string pitch_str {get_node_value(lat_node, "pitch")};
+  std::vector<std::string> pitch_words {split(pitch_str)};
+  if (pitch_words.size() != dimension_words.size()) {
+    fatal_error("Number of entries on <pitch> must be the same as the "
+                "number of entries on <dimension>.");
+  }
+  pitch[0] = stod(pitch_words[0]);
+  pitch[1] = stod(pitch_words[1]);
+  if (is_3d) {pitch[2] = stod(pitch_words[2]);}
+
+  // Read the universes and make sure the correct number was specified.
+  std::string univ_str {get_node_value(lat_node, "universes")};
+  std::vector<std::string> univ_words {split(univ_str)};
+  if (univ_words.size() != nx*ny*nz) {
+    std::stringstream err_msg;
+    err_msg << "Expected " << nx*ny*nz
+            << " universes for a rectangular lattice of size "
+            << nx << "x" << ny << "x" << nz << " but " << univ_words.size()
+            << " were specified.";
+    fatal_error(err_msg);
+  }
+
+  // Parse the universes.
+  universes.resize(nx*ny*nz, C_NONE);
+  for (int iz = 0; iz < nz; iz++) {
+    for (int iy = ny-1; iy > -1; iy--) {
+      for (int ix = 0; ix < nx; ix++) {
+        int indx1 = nx*ny*iz + nx*(ny-iy-1) + ix;
+        int indx2 = nx*ny*iz + nx*iy + ix;
+        universes[indx1] = stoi(univ_words[indx2]);
+      }
+    }
+  }
+}
+
+//==============================================================================
+
+int32_t&
+RectLattice::operator[](const int i_xyz[3])
```

Comment:
Will this work for a 2D lattice? (same question goes for hexagonal lattice)

Reply:
It will as long as `i_xyz[2] = 0` which will always be returned by `get_indices` for 2D lattices.